### ISSUE TITLE
Failed to get token

### ISSUE THREAD

**Issue by higher-faster-better** [omise/omise-dotnet#19]
Hi.

I've tried this in Xamarin.Android.
[https://github.com/omise/omise-dotnet/blob/master/Omise.Examples/Example.cs](url)

It takes long time to get response and finally it says "A task was canceled."
so I can't get token.

What is the issue?

**Comment by higher-faster-better** [comment]
Hey.

I wanted to get help to fix this issue.
Isn't support available to help?

**Comment by robinclart** [comment]
Hi @higher-faster-better. Robin, Omise CTO here. This was a long weekend here in Thailand, since today is a national holiday. And we have no developers available until tomorrow morning to look into this. But rest assured we took notice and someone will help you resolve this tomorrow. Have a nice evening.

**Comment by chakrit** [comment]
@higher-faster-better Hi, sorry for slow reply. It has been a super busy week.

"A task was canceled." is a very generic error. Can you try to get a more specific exception message? One way to do this is to save the `Task` object when running and check the `task.Exception` property like this:

```csharp
var t = Retrieve_Retrieve(); // no await, use manual wait()
t.Wait();

var ex = t.Exception; // <-- or set a breakpoint here and inspect it.
Console.WriteLine($"exception: {ex.Message}");
Console.WriteLine($"inner exception: {ex.InnerException.Message}");
```

**Comment by chakrit** [comment]
@higher-faster-better I have made a quick throwaway app to test and it looks like it is working correctly: https://github.com/chakrit/omise-xamarin-forms 

![2017-07-11 at 10 06 pm](https://user-images.githubusercontent.com/45545/28075304-47fb45a2-6685-11e7-9484-489121aff5b0.png)

I'll try create a more complicated example later in the week.

**Comment by higher-faster-better** [comment]
Hi, Thanks for your reply.
It works good for IOS. It doesn't work only on Android.

**Comment by higher-faster-better** [comment]
Please try your project on Android and check it if it works.
Thanks.

**Comment by higher-faster-better** [comment]
Here is the exception.

exception: A task was canceled.
[Choreographer] Skipped 432 frames!  The application may be doing too much work on its main thread.
[CubicBezierInterpolator] CubicBezierInterpolator  mControlPoint1x = 0.15, mControlPoint1y = 0.7, mControlPoint2x = 0.2, mControlPoint2y = 0.98
[CubicBezierInterpolator] CubicBezierInterpolator  mControlPoint1x = 0.6, mControlPoint1y = 0.9, mControlPoint2x = 0.8, mControlPoint2y = 1.0
[CubicBezierInterpolator] CubicBezierInterpolator  mControlPoint1x = 0.15, mControlPoint1y = 0.7, mControlPoint2x = 0.2, mControlPoint2y = 0.98
[CubicBezierInterpolator] CubicBezierInterpolator  mControlPoint1x = 0.6, mControlPoint1y = 0.9, mControlPoint2x = 0.8, mControlPoint2y = 1.0
[CubicBezierInterpolator] CubicBezierInterpolator  mControlPoint1x = 0.15, mControlPoint1y = 0.7, mControlPoint2x = 0.2, mControlPoint2y = 0.98
[CubicBezierInterpolator] CubicBezierInterpolator  mControlPoint1x = 0.6, mControlPoint1y = 0.9, mControlPoint2x = 0.8, mControlPoint2y = 1.0
[CubicBezierInterpolator] CubicBezierInterpolator  mControlPoint1x = 0.15, mControlPoint1y = 0.7, mControlPoint2x = 0.2, mControlPoint2y = 0.98
[CubicBezierInterpolator] CubicBezierInterpolator  mControlPoint1x = 0.6, mControlPoint1y = 0.9, mControlPoint2x = 0.8, mControlPoint2y = 1.0
[mali_winsys] new_window_surface returns 0x3000

**Comment by higher-faster-better** [comment]
Hey.
Any updates? I am still waiting for your support to complete integration.

**Comment by chakrit** [comment]
@higher-faster-better Hi, from the error message it looks like you may be doing network calls on the main thread:

```
[Choreographer] Skipped 432 frames! The application may be doing too much work on its main thread
```

Can you make it do that in the background instead, maybe using `Executor` or some other Android-provided functionality.

I am setting up an android test project right now, but this may take quite a while.

**Comment by chakrit** [comment]
@higher-faster-better Hi I've done some further tests and it looks like `async/await` plus the Xamarin-managed HTTP clients are really really slow and will often crash the thread that's doing the call outright.

I have tried changing from Xamarin HTTP client to use android-native one and added the TLS 1.2 support (required by our HTTP API) and this seems to work much more reliably:
 
![android](https://user-images.githubusercontent.com/45545/28207519-327f40e4-68b4-11e7-932d-4c902077da14.png)

Can you try this settings ☝️ and let me know how that goes?

**Comment by higher-faster-better** [comment]
Thanks for your reply.
I've tried your solution and it works!
But it's not perfect solution because it AndroidClientHandler requires Android 5.0 or later.
It means that I can't use this payment gateway in lower Android version.
(Actually I tested it in Android 4.4 and doesn't work.)

The wired thing is that only getting token doesn't work on my app.
I implemented creating customer and it works good.
And I was able to success in doing payment by card with token that I created in web.

Did you notice that?
Why does NOT only Getting Token work?

Look forward to hearing your opinion soon.
Thanks.

**Comment by chakrit** [comment]
@higher-faster-better Yes, we are aware of the limitation – this was due to our higher PCI-DSS requirements on our token storage server. Token creations go through our secured vault which requires strict TLS 1.2 support. It is not available below android 4.4 (or requires nasty hacks). I am not aware of the requirements of  `AndroidClientHelper`.

However, if you use our native [Android SDK](https://github.com/omise/omise-android) it should work at 4.4 and a few minor versions below. (cc @pitiphong-p to confirm)

I'd recommend not going below 4.4, it's better for your application's security anyway. An alternative would be to embed a custom TLS library from Nuget but I'm not sure if there is a good one available and it'd effect regular server-side .NET users too.
